### PR TITLE
 Dialog component

### PR DESCRIPTION
**Pull request by mateusmarquezini**
## Hey, I just made a Pull Request!

fix: #3497

Created a new example of how to use the Dialog component. Below, some screenshots examples of it.
I was wondering about the best way to explain how to use it, so I put it into the Dialog.

Normal use:
demo_1

With scroll:
demo_6_with_scroll

Dark mode:
demo_2_dark_mode

screen size / 1440
demo_4_1440

screen size / 768
demo_3_768

screen size / 520
demo_5_520

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/warm-hotels-happen.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'storybook': minor
+---
+
+Created a new example of how to use the Dialog component in Storybook
```
Comment (Fox32) [inline .changeset/warm-hotels-happen.md:5]: As it's just a Storybook story, you don't need to add a changeset. This change won't be noticeable to package consumers.

Comment (mateusmarquezini) [reply .changeset/warm-hotels-happen.md]: ok, thanks! :)

**Comment by Fox32** [comment]
I know that Backstage has its own style and isn't following Material Design closely and Material Design is only a base on which we can do our own design. [They seem to use the](https://material.io/components/dialogs#actions) [text style for dialog buttons](https://material.io/components/buttons#text-button) like Backstage is already doing in some dialogs. 

For example here:
image

However the suggested story is using the elevated/contained button style:

image

(Same goes for the alignment)

I wonder if we should select a different margin if we decide to go with contained buttons as the dialog looks a bit cut off at the bottom.

**Comment by Fox32** [comment]
Another thing not directly related: I lately thought about whether I want to use a drawer or a dialog for certain things. I like that the dialog story explains the usage, maybe we can have the same for the drawer? It's a bit hidden dialog, but I really like that part, but it serves as a nice example content. Maybe you could put it both inside and outside of the dialog (not copy and paste but by defining a separate component)?

**Comment by adamdmharvey** [comment]
@Fox32 , I echo your comments as I was just modifying a dialog yesterday and noticed all of the buttons right aligned as well. (different than this story)

But @mateusmarquezini 's story directly aligns to the Figma outline for "Forms" here:
https://www.figma.com/community/file/850673348101741100
(select Form Validation at top left)

I guess what is interesting is how to bring along a dev through the strategy path, is it figma first, then storybook for the raw copy/paste elements, etc, or is Figma the rough placeholder and storybook becomes the FINAL location. (as some of the wording added here is I can see a direct copy from the Figma story) I wonder if the Spotify design team may have some input there.

But exactly as you said, if this story is merged as is, it therefore suggests there is other work to then refactor existing dialogs to match it. We also have that tight rope to walk with making a consistent design easy to use for devs - with ideally as little customization as possible _needed_ (e.g., having to always remember to add the "root.classesLeftNav" every time someone adds a Dialog per straight Material-UI, vs. a customized "pre-Backstage themed" component that is easier to use), but potentially allowing wild customization where it makes sense (hopefully less often, and more likely in plugins with raw Material-UI?).

**Comment by freben** [comment]
I think we can at least ignore Google's Material Design guidelines for the purposes of this discussion. For practical reasons we'll rely on material-ui's implementation of it as the "base plate", leaning into and benefiting from their efforts.

From a plugin maker's perspective, we need to make it trivial to do the right thing. If we want a certain behavior on dialogs, we shall avoid adding process for specifying certain props, or styles, or even classes. We shall instead focus on getting central base theme overrides in place so it gets applied consistently. [Here](https://github.com/backstage/backstage/blob/master/packages/theme/src/baseTheme.ts#L253)'s an example where I did an addition like that recently, to better adapt cards to css-grid layouts.

But at the end of the day, I personally am pretty adamant that we should leave MUI alone to the greatest extent possible. If MUI places buttons on one side of dialogs per default, I feel we better have a pretty darn good reason to change that. One that's even significantly better than the Material Design guidelines that led to it in the first place, essentially.

Ok that's me rambling a bit. Now over to somebody that actually knows design / UX stuff which I frankly don't. :)

**Comment by freben** [comment]
Oh to clarify @mateusmarquezini this is not at all a stab at you. This was just me musing a bit about how I hope the project could in general approach component design - "raw MUI first". @adamdmharvey puts it nicely, in terms of where to look for authoritative information about design choices.

I hope that the answer will normally be "the MUI official docs", and https://backstage.io/storybook. If one wants to contribute a new shared component to the core, it should be nicely and reusable composed from MUI primitives and tweaked as little as possible so that it's more easily migrated to newer MUI releases down the line.

**Comment by mateusmarquezini** [comment]
> Oh to clarify @mateusmarquezini this is not at all a stab at you. This was just me musing a bit about how I hope the project could in general approach component design - "raw MUI first". @adamdmharvey puts it nicely, in terms of where to look for authoritative information about design choices.
> 
> I hope that the answer will normally be "the MUI official docs", and https://backstage.io/storybook. If one wants to contribute a new shared component to the core, it should be nicely and reusable composed from MUI primitives and tweaked as little as possible so that it's more easily migrated to newer MUI releases down the line.

Humm, ok thanks for clarifying it @freben! I agree about make it less tweaked as little as possible, even because it would turn a little bit hard to reuse and maintain the design of the components.

As the best approach here is to make it more easily to reuse and contributing using shared components, and tweak them as little as possible, I understand that it's better here to consider the buttons on DialogActions aligned to the right without the tweak overriding the justifyContent prop as I did, as @Fox32 mentioned here, right?

image

**Comment by Fox32** [comment]
I also don't want to say that we have to follow material design closely. It's like bootstrap, if everyone is using it all UIs look the same. And this is actual a point of diversification of the design. My comment is more about how we enforce this design (on a technical level).

If we follow the same approach we did for other component, we could also export a `DialogActions` that provides that styling? But that still has the issue that users have to use the correct import...

**Comment by freben** [comment]
No need i think, there's theme override support for the builtin mui `DialogActions`. https://material-ui.com/api/dialog-actions/#css

**Comment by benjdlambert** [comment]
@katz95 is this something that you could update in figma? We ideally want to follow the MUI guidelines unless there's a super good reason not to 👍

**Comment by katz95** [comment]
Yes, we can totally update this on our end to follow MUI guidelines. Thanks you all!

**Comment by mateusmarquezini** [comment]
Ok, done!

Screen Shot 2021-04-02 at 08 05 58

**Review by Rugvip — APPROVED** [review]
Not merging yet to allow for some more 👀 , but I say let's :shipit: because I think the code in the story is what we want. If we want the dialog to look different then I think we should handle the rest elsewhere

**Pull request merged**